File: JS/portfolio.js
import { Money } from './money.js'

export class Portfolio {
    constructor() {
        this.moneys = [];
    }

    add(...money) {
        this.moneys = this.moneys.concat(money);
    }

    evaluate(currency) {
        let total = this.moneys.reduce( (sum, money) => {
            return sum + this.convert(money, currency);
        }, 0);
        return new Money(total, currency);
    }
    convert(money, currency) {
        let exchangeRates = new Map();
        exchangeRates.set("EUR->USD",1.2);
        exchangeRates.set("USD->KRW",1100);

        let key = money.currency + "->" + currency;
        if(money.currency === currency)
            return money.amount;
            
        return money.amount * exchangeRates.get(key);
    }
}

File: PY/portfolio.py
import functools
import operator
from money import Money

class Portfolio:
    def __init__(self) -> None:
        self.moneys = [];

    def add(self, *moneys):
        self.moneys.extend(moneys)

    def evaluate(self, currency):
        total = 0.0
        failures = []
        for m in self.moneys:
            try:
                total += self.__convert(m, currency)
            except KeyError as ke:
                failures.append(ke)

        if len(failures) == 0:
            return Money(total,currency)
        
        failureMessage = ",".join(f.args[0] for f in failures)
        raise Exception("Missing exchange rate(s):[" + failureMessage + "]")

    def __convert(self, money, currency):
        exchangeRates = {'EUR->USD': 1.2, "USD->KRW": 1100}
        key = money.currency + '->' + currency
        if money.currency == currency:
            return money.amount
        return money.amount * exchangeRates[key];
